Sources/StreamChatUI/ChatMessageList/ChatMessageListVC.swift

```diff
@@ -533,3 +606,103 @@ open class ChatMessageListVC:
         !(touch.view is UIControl)
     }
 }
+
+// MARK: - Backwards Compatibility DataSource Diffing
+
+@available(iOS 13.0, *)
+internal extension ChatMessageListVC {
+    /// Setup the `UITableViewDiffableDataSource`.
+    func setupDiffableDataSource(for listView: ChatMessageListView) {
+        let diffableDataSource = UITableViewDiffableDataSource<Int, ChatMessage>(
+            tableView: listView
+        ) { [weak self] _, indexPath, _ -> UITableViewCell? in
+            /// Re-use old `cellForRowAt` to maintain customer's customisations.
+            let cell = self?.tableView(listView, cellForRowAt: indexPath)
+            return cell
+        }
+
+        self.diffableDataSource = diffableDataSource
+        listView.dataSource = diffableDataSource
+
+        /// Populate the Initial messages data.
+        var snapshot = NSDiffableDataSourceSnapshot<Int, ChatMessage>()
+        snapshot.appendSections([0])
+        snapshot.appendItems(dataSource?.messages ?? [], toSection: 0)
+        diffableDataSource.apply(snapshot, animatingDifferences: false)
+    }
+
+    /// Transforms an array of changes to a diffable data source snapshot.
+    func updateMessagesSnapshot(with changes: [ListChange<ChatMessage>], completion: (() -> Void)?) {
```

Comment:
Unfortunately, if we have inconsistent updates it will still crash. The hope here is that applying the snapshots has better robustness in terms of concurrency.